### ISSUE TITLE
How to improve the efficiency of sort in CrypTen?

### ISSUE THREAD

**Issue by amyzx** [facebookresearch/CrypTen#337]
Hi, sorry to bother you again. 

I implement a bitonic sort function in CrypTen, which can sort a data array and its corresponding idx array.

I find this algorithm is inefficient. So I try to use multiprocessing to improve this sort algorithm.  However, this program seems to be stuck with **pool.starmap** function. 

The following is my [code](https://github.com/amyzx/crypten_learn/blob/main/sort_enc.py). I am not familiar with multiprocess programming in crypten. Could you give me some suggestions?

Looking forward to your reply. Thanks a lot!

**Comment by knottb** [comment]
If you are asking about the deadlock in your multiprocess Pool, I think you are better off removing that. I believe PyTorch will parallelize this under the hood based on system parameters.

I haven't experimented much with using the `multiprocess` library external to CrypTen while using CrypTen objects. I imagine a number of issues could arise that lead to timeouts. The first that comes to mind is we are using `torch.distributed` to communicate between processes. This connection is setup inside of `crypten.init()`. Additionally the `@mpc.run_multiprocess` decorator also initializes the connections between processes. If you use multiprocessing within the code block, it is possible that these parties do not have the ability to communicate with each other and hang during MPI calls.

To summarize, I haven't experimented much with nested multiprocessing like this, but I can imagine it could lead to some complex issues to debug. I think the parallelization you want may be implemented at the PyTorch level under the hood so these may already heavily exploit parallelization.

**Comment by amyzx** [comment]
Thanks a lot!
